docs/procbot-architecture.md

```diff
@@ -0,0 +1,322 @@
+
+```
+123:
+ABCDEFGHIJKLMN
+ABCDEF
+ABCDEFGHIJKLMNOPQRSTUVWXYZ
+
+456:
+1234
+1234567890
+12
+```
+
+These are now out of order, and not the desired output.
+
+Obviously this has potentially disasterous consquences depending on the data being processed. However, similarly, we do not want to wait for the completion of events that are not related as this slows execution.
+
+Workers are designed to solve the concurrency issue presented here by allowing a ServiceListener to define a context for the events that they receive. Events that are related are serialised, ensuring that any processing and emitting is carried out completely for an event before the next event in the context is operated upon, but ensuring that separate events can also be processed asynchronously to achieve a suitable throughput.
+
+Workers and WorkerClient Bots are instrumental to achieving event handling based on context.
+
+Whenever a ServiceListener receives a new event from an external source, it uses methods in the WorkerClient to queue the event into the relevant context. This is carried out by the ServiceListener implementing a method that creates a new Worker should a relevant context not already exist, or retrieving a current Worker should a Worker for the context already exist. The WorkerClient then adds the passed event to that Worker. WorkerClient Bots hold a map of each Worker along with the value of its context.
+
+The Worker class itself stores, in order of reception, all current events that have been queued based on a specific context. This context is a generic type and could be anything; a string, a number, an object, etc.
+
+Workers are self-regulating; every time an event has been processed by a Client Bot the Worker will check to see if there are anymore events to pass to a Client. It will do this until the queue is empty and then remove itself from the map of Workers held in the WorkerClient.
+
+ServiceListeners all rely on the WorkerClient class to perform the context duties for queueing new events.
+
+### ServiceEmitters
+
+ServiceEmitters act as a Client Bots ability to send data to external services. This may be a simple 'dumb' fire and forget, or the act of using an API to request data and receive responses.
+
+Every ServiceEmitter expects a `ServiceEmitRequest` object which holds data allowing it to communicate relevant to the external service it acts as an arbiter for. As a `ServiceEmitRequest` object could potentially be passed to many ServiceEmitters, they contain a `contexts` property, which should contain a unique property named by the service that the Client Bot wishes to use, the value of which is an object that allows a service to make requests to that service.
```

Comment:
Good spot. Yep, the `ServiceEmitRequest` has the `contexts` property. I need to rewrite that sentence to make sense!

As to the question, I had considered only instantiating one instance of a `Service[Listener/Emitter]` and then sharing it between all bots that want one. However, I decided against this and went for new instances for each Bot for a couple of reasons:
1. It makes each Bot self-contained. If, for some reason, a particular Service died because of something a bot listened for/emitted, it won't take down out any other data another Bot might also be passing through it at the time.
2. It makes handling of sub-service infrastructure (such as Express usage) a lot easier to handle.

Happy to discuss the pros and cons of this if there's strong opinion!
I'm happy to discuss